diff -urN linux-2.6.9-rc2.org/lib/Kconfig linux-2.6.9-rc2/lib/Kconfig
--- linux-2.6.9-rc2.org/lib/Kconfig	2004-09-16 22:15:16.000000000 +0200
+++ linux-2.6.9-rc2/lib/Kconfig	2004-10-05 23:37:40.000000000 +0200
@@ -39,5 +39,23 @@
 config ZLIB_DEFLATE
 	tristate
 
+#
+# reed solomon support is select'ed if needed
+#
+config REED_SOLOMON
+	tristate
+	
+config REED_SOLOMON_ENC8
+	boolean
+
+config REED_SOLOMON_DEC8
+	boolean
+
+config REED_SOLOMON_ENC16
+	boolean
+
+config REED_SOLOMON_DEC16
+	boolean
+
 endmenu
 
